### ISSUE TITLE
[FEA] Support `duplicate_find_option::FIND_ALL` for `lists::index_of`

### ISSUE THREAD

**Issue by ttnghia** [rapidsai/cudf#12212]
Currently, the API `lists::index_of` can only search for the first or last index of the given key in the lists. In some cases, we need to search for all the positions where the key appears in the search space (lists). For example, we want to remove an element from the lists:
```
remove({1, 2, 3, 4, 5}, 3) ==> {1, 2, 4, 5}
```

In order to implement such API (`remove`), we need to support `duplicate_find_option::FIND_ALL` for `lists::index_of` so we can search for all indices of the search key as a gather map and generate the output using that gather map.

**Comment by GregoryKimball** [comment]
Thanks @ttnghia for raising this issue. Would the "find_all" behavior in [parsing_utils.cu](https://github.com/rapidsai/cudf/blob/e67b94b19056282325bd5bad823fd66aeafd92c5/cpp/src/io/utilities/parsing_utils.cu#L117) be useful for this new option?

**Comment by ttnghia** [comment]
No, it is similar but totally different here. That `parsing_utils.cu#find_all_from_set()` is dealing only with `char` type. For general lists (which may be nested) then the implementation is very different.
